### PR TITLE
mlang Has for java

### PR DESCRIPTION
**Pull request by jlhughes**
Need to discuss if we should test for empty string as well as null.

### CODE REVIEW

File: src/foam/mlang/mlangJava.js [hunk 1]
```diff
@@ -323,6 +323,17 @@ foam.CLASS({
 });
 
 
+foam.CLASS({
+  refines: 'foam.mlang.predicate.Has',
+
+  methods: [
+    {
+      name: 'f',
+      javaCode: 'return getArg1().f(obj) != null && getArg1().f(obj) != "";'
```
Comment (adamvy-google) [inline src/foam/mlang/mlangJava.js]: So JS checks for undefined, null, empty string and empty array.

I'm not sure what the correct behaviour is, but we should be consistent in any case.

Comment (jlhughes) [reply src/foam/mlang/mlangJava.js]: Testing for an empty array is not straight forward.  Testing for an array can be accomplished with 'value.getClass().isArray()' or 'instanceof _Type_[]', but then to get the length, it needs to be cast or you have to know the _Type_.

**Comment by jlhughes** [comment]
Hi  Adam
The test for empty string and empty Array was added some time ago, appears my comment regarding it was hidden. 
Joel

File: src/foam/mlang/mlangJava.js [hunk 2]
```diff
@@ -323,6 +323,20 @@ foam.CLASS({
...
 
 
+foam.CLASS({
+  refines: 'foam.mlang.predicate.Has',
+
+  methods: [
+    {
+      name: 'f',
+      javaCode: 'Object value = getArg1().f(obj); '+
+        'return ! (value == null || '+
```
Comment (adamvy) [inline src/foam/mlang/mlangJava.js:333]: Are you sure this compiles?

I don't believe isEmpty() is a method on all Objects in java.

I also don't think we should compare the output of toString() as that may require building strings we don't need which is slow.

Comment (jlhughes) [reply src/foam/mlang/mlangJava.js:333]: yes, I commented on this previously, but it was lost/hidden:
"Testing for an empty array is not straight forward. Testing for an array can be accomplished with 'value.getClass().isArray()' or 'instanceof Type[]', but then to get the length, it needs to be cast or you have to know the Type."
How would you like proceed?

Comment (adamvy) [reply src/foam/mlang/mlangJava.js:333]: We can use getClass().isArray() && java.lang.reflect.Array.getLength(value)

**https://docs.oracle.com/javase/7/docs/api/java/lang/reflect/Array.html#getLength(java.lang.Object)**

**Comment by jlhughes** [comment]
Will re-open on new fork.
